### ISSUE TITLE
Wrong formatting of Tango attributes

### ISSUE THREAD

**Issue by LJBD** [taurus-org/taurus#392]
Hello!

Taurus 3.x used to read format from attribute configuration in Tango and display it accordingly. Taurus 4.0.x does not do that.
I wrote 1.123456789 to float_scalar attribute of a TangoSchemeTest device. The configured format is %6.2f. You can see behaviour of Taurus 3.7.3, 4.0.1, 4.0.3 and 4.0.4-alpha (I used Carlos Falcon's bug-387 branch).
![taurus_display_format_test](https://cloud.githubusercontent.com/assets/9516795/22457124/664eba5c-e796-11e6-8df5-abb8064db597.JPG)

**Comment by cmft** [comment]
I can reproduce it. Thanks!

**Comment by cpascual** [comment]
If I remember correctly, we left this out of Taurus4 knowingly because doing a proper (scheme-agnostic) support for formatting in taurus required some deep discussion. I think that we could now open such discussion (maybe it is even worth a TEP).

See https://github.com/taurus-org/taurus/pull/393#issuecomment-276603663:

>We should have a scheme-agnostic property for specifying the format (I propose the "new" standard python string-formating better than old style c-like formatting because it would integrate more naturally with quantities), and encode/decode it from whatever an scheme uses.

**Comment by cmft** [comment]
Hi @cpascual , @LJBD ,
I think we can fix it, with minor API changes (as temporally solution) and assuming that taurus  uses the "new" standard python string-formatting. The only scheme affected for the moment is Tango.  I implemented a solution in one of my branches https://github.com/cmft/taurus/tree/bug-392.

What do you think? Is it enough?

**Comment by LJBD** [comment]
It looks all right from a user point of view. I don't feel versed in taurus enough to be able to asses if the code is written well, though.
Anyway, I've noticed a funny behaviour: I've launched a ATK Attribute property editor and a TaurusLabel for the same attribute. When I'm changing format and clicking "Apply changes", the TaurusLabel updates instantly. When I'm changing units, it does not - it updates units only when value of the attribute changes. Is that an expected behaviour?

**Comment by cpascual** [comment]
> What do you think? Is it enough?

I did some comments on the commits of [your branch]( https://github.com/cmft/taurus/tree/bug-392), but I think that we should stop and discuss this more before actually looking into implementations

Here are some points that I think should be clarified:

- Does a "format" really belong to the Model? IMHO it is clearly something related to the view and not the Model (e.g. 2 different widgets may require different formats). 
- Should this be a scheme-specific solution or a scheme-agnostic one? (e.g., I think that epics also defines some format).
- If we implement a `TaurusAttribute.format` property, what would be its meaning? Just a hint? 
- In practice, it seems that the main use-case for the format is to hint on the _precision_ of the value (i.e. hinting the number of significant significant digits that the view should use). Perhaps instead of "format" (which belongs to the View) we should consider having a "precision" property (which is indeed something more adequate for the Model). Then each view could take the precision into account for deciding its own "format" (maybe the precision could be codified using a syntax inspired in that of the c-like string formatting, without the format placeholder)

**Comment by cpascual** [comment]
>  it updates units only when value of the attribute changes. Is that an expected behaviour?

I would say it is "expected" for me... but "expected" doesn't necessarily mean "correct".

Note that in Taurus, the unit is not a "configuration" of the attribute as a whole, but something intrinsic to a *value* (note that in **Taurus** the `rvalue`, the `wvalue` and the limits may all have different units, and it is just a particularity of the *Tango scheme* that they are all the same)

**Comment by cpascual** [comment]
BTW, I just checked, and epics actually defines "precision" instead of "format". I think we could use the same conventions: http://cars9.uchicago.edu/software/python/pyepics3/pv.html#string-representation-for-a-pv

**Comment by cpascual** [comment]
BTW, I just checked, and pyepics actually defines "precision" instead of "format". I think we could use the same conventions to transform precision into format :
http://cars9.uchicago.edu/software/python/pyepics3/pv.html#string-representation-for-a-pv

**Comment by cpascual** [comment]
BTW, I just checked, and pyepics actually defines "precision" instead of "format". I think we could use the same conventions to transform precision into format :
http://cars9.uchicago.edu/software/python/pyepics3/pv.html#string-representation-for-a-pv

**Comment by cpascual** [comment]
So, to summarize:

I propose the following:

- we add a "precision" attribute to the TaurusAttribute class with the same meaning as in pyEpics ("number of decimal places of precision for floating point values")
- We encode/decode it for each scheme (e.g. for tango, we would have:  format="%6.2f"  <--> precision=2 )
- We make the display widgets use this precision as a hint for display

Any volunteers for a PR?

**Comment by sergirubio** [comment]
Hi, 

In my opinion, "precision" is much more limited than format; and cannot be use to specify a hex number or whether to fill with '0' the left side of the number. 

As we also have the need to represent binary or hex numbers in taurus I think it would be useful to keep the "format" attribute as versatile as it is in the '%' python operator. I think that it would be much more pythonic that just a precision  (that you would have to adapt to a format string anyway).

I rather prefer epics "precision" to be adapted to a "format" string than the opposite.

Sergi

**Comment by cpascual** [comment]
Hi @sergirubio , I agree that "format" would allow more customization of displays than `precision`, but I prefer the precision for the following reasons:

- In taurus, it should support quantity objects (and enumerations, and numpy arrays, etc). For this it should be based on the [new string formatting style](https://docs.python.org/2/library/string.html#format-string-syntax) rather than the c-style one. This makes it already incompatible with tango (so encode-decode would still be needed).
- "format" is something that belongs to the View in the M-V-C pattern. The TaurusAttribute is a Model. Different Views (widgets) will want to use different formats for the same model. This is IMHO a huge design-flaw Tango, and I think that Epics does it better by using `precision`. The precision is something intrinsic to the value: it communicates how many of the digits are significant... this affects formatting, but also other stuff such as uncertainties estimation. Regarding the format of the display string, if  an attribute declares its precision, a widget may choose to format the display to use the maximum precision... or to display less than that.
- The format customization can still be implemented at a generic widget level (e.g. in  `TaurusBaseComponent.displayValue`). And it can be implemented to make use of the Model `precision` by default.

**Comment by cmft** [comment]
@cpascual The precision is a nice concept, and should be nice to have it, but in Tango scheme it will affect to the stored value (rvalue and wvalue).  Assuming that we don't strongly link `format` (Tango's concept) and `precision` (Taurus' concept), and it can be configured at TaurusAttribute level, if  we use the same attribute in several GUIs the showed value will be "limited" by the minor precision, we will see precision problems due the events (from different clients).

IMHO, we should have a fragment `#format`,  and all Tango model should include it by default (to imitate T3).

**Comment by cpascual** [comment]
>  Assuming that we don't strongly link format (Tango's concept) and precision (Taurus' concept), and it can be configured at TaurusAttribute level, if we use the same attribute in several GUIs the showed value will be "limited" by the minor precision, we will see precision problems due the events (from different clients).

I am not sure, but I think that you are confusing the "displayed precision" (which is actually a "format") with the model precision. A normal widget (not a config widget) should not change the model precision (just as it should not change the Tango units or the Tango format) because that may have unintended effects on other clients. Normal widgets can use whatever format or precision or units they prefer **for display** without changing the model ones.

**Comment by cpascual** [comment]
> IMHO, we should have a fragment #format

I also think that we could use the fragment to pass some formatting preferences to the View (widgets)
Note that the fragment does not get to the model, and stays in the view.
The only thing is that we should design something compatible with the current concept of fragment.

**Comment by cmft** [comment]
> I am not sure, but I think that you are confusing the "displayed precision" (which is actually a "format") with the model precision. A normal widget (not a config widget) should not change the model precision (just as it should not change the Tango units or the Tango format) because that may have unintended effects on other clients. Normal widgets can use whatever format or precision or units they prefer for display without changing the model ones.


Maybe I 	expressed badly, I put  an example.
If a tango attribute has precision 1, it can generate (rvalues, wvalue) values like 1.1, 2.2,... and if the precision is 2 it generates values like 1.11 2.22,... I'm right?

If yes, if two clients defining the same tango attribute with different precision (as I put in the example). The client who has the model with precision 2 will show values like 1.11 or 1.10 depending which client write the value (due to the client with precision 1 only can write 1.1). I mean, the precision (as concept) does not change for any of the clients but the value resolution is truncated by the client with less precision since the values are saved (in the db) with the corresponding precision.

For that reason, I prefer to use only "format" as a view option if we allow that the same model in different clients can set different precision.

**Comment by cpascual** [comment]
No. The `precision` would not affect at all at the rvalue or wvalue. It may affect at how those are *displayed* but not at their values at all.
Even if you consider the case of an input widget such as the TaurusWheelEdit, which may make use of the precision to decide how many digits it allows the user to set, this would not change anything, because:
a) the widget could (should) provide a way for the user to change the number of digits (locally at the widget, not at the model)
b) even if it doesn't, that would only mean that the written values would be rounded at the given decimals, which is the same that happens in taurus3 if you write 1.23 on a TaurusLineEdit that is editing a double with much more precision.

**Comment by cmft** [comment]
> No. The precision would not affect at all at the rvalue or wvalue. It may affect at how those are displayed but not at their values at all.


Ok. I was misunderstanding the concept. So, at the end the effect of use `precision` or the `format` (T3) would be the same. The use of `precision` is due to the meaning of the concept (attribute name) cause both attributes affect the View. 

Are you planning to link  the tangoattribute `format` and the tangoattribute `precision`?

IMHO, it should be a View option not for the Model

**Comment by cpascual** [comment]
> So, at the end the effect of use precision or the format (T3) would be the same. 
In *what respects to displaying* yes, they would act similar. 

> The use of precision is due to the meaning of the concept (attribute name) cause both attributes affect the View.

Not exactly. As I explained before, precision is a Model-related concept which can be used for various things. Hinting a format for display in Views is one of them.

> Are you planning to link the tangoattribute format and the tangoattribute precision?

IMHO, PyTango's format should be decoded to get taurus' precision. I do not think that we should ever set PyTango's format from a Taurus widget (unless it is a Tango-specific config widget talking straight to PyTango)
 
> IMHO, it should be a View option not for the Model

If you talk about the format, I totally agree. `TaurusBaseComponent` could define a `displayFormat` property, which could get a default value from its model's `precision`.
And, I would make this "displayFormat" use the new string formatting style, so that you can do things such as selecting long or short unit names

**Comment by cmft** [comment]
In an internal discussion between @cpascual and me, we discarded the option to add new attribute to the model, and we agree to solve the problem at "View level". We see two possibles solutions:

- The best option for us is delegate the work to the widget.  This will allow to set the formatting via setter (like the compact mode). This solution allows to specify different formatting for the same attribute in each widget, but the connection with the TangoAttribute format attribute (that exists in Taurus3) will disappear. 
- The second option is to allow specify the format via URI, i.e. give abusively the string-formatting using a `#fragment`.

e.g.
```
# To use the Tango format
sys/tg_test/1/float_scalar#format
# To forze a format
sys/tg_test/1/float_scalar#format#format={:.2}
```

I think that both solution could allow to set the configuration via GUI and programmatically, although the second solution in a less intuitive way.

What do you think?

NOTE: In both solutions the new python string formatting must be used.

**Comment by sergirubio** [comment]
Whatever solution you choose we need a way to apply the Tango format to all TaurusLabel declared in already existing applications.

This could be easily solved if we can provide a TaurusLabel.setDefaultFormatter class method that allows to set the default label format method for the whole application. That method could be a taurus.core.tango.utils.tangoFormatter(model,value) method capable of getting (and caching) attribute formats from Tango.

My point is just adding a line like this in our gui's to have Tango format as default:

  TaurusLabel.setDefaultFormatter(tauru.core.tango.utils.tangoFormatter)

Format could be overriden later for each widget, but at least we would have a one-line migration solution.

Sergi

**Comment by sergirubio** [comment]
Just one point regarding the "new" (already 10 years old) python string formatting.

It's nice, but only used in python. The "old" method using "%" is more popular because it's familiar for any user or developer on any language, not only python.

Maybe it's not pythonic, but for usability I would rather use the old format. Anyway, if we add a hook for setting "formatter" methods we can leave this decision to the developer.

**Comment by cpascual** [comment]
> This could be easily solved if we can provide a TaurusLabel.setDefaultFormatter class method (...)

I need to consider this more carefully, but on first sight I am not convinced about the implementation details (maybe we can have an in-person discussion)

> Maybe it's not pythonic, but for usability I would rather use the old format.

But the reason for preferring the "new" one is that with the old one you cannot manage [quantity formatting](http://pint.readthedocs.io/en/0.6/tutorial.html#string-formatting).

**Comment by cpascual** [comment]
Please have a look at #444 which proposes a fix for this

**Comment by cpascual** [comment]
Fixed by the merge of #444
